### PR TITLE
Fixed #32178 -- Allowed database backends to skip tests and mark expected failures.

### PR DESCRIPTION
**Pull request by hramezani**
https://code.djangoproject.com/ticket/32178

### CODE REVIEW

File: django/db/backends/base/creation.py [hunk 1]
```diff
@@ -47,6 +67,8 @@ def create_test_db(self, verbosity=1, autoclobber=False, serialize=True, keepdb=
                 self._get_database_display_str(verbosity, test_database_name),
             ))
 
+        self.mark_tests()
```
Comment (timgraham) [inline django/db/backends/base/creation.py:70]: This should only be called when the Django test suite is running. Perhaps with an environment variable which `runtests.py` could set, unless we can come up with a more elegant solution.

File: django/db/backends/base/creation.py [hunk 2]
```diff
@@ -27,6 +32,21 @@ def _nodb_cursor(self):
     def log(self, msg):
         sys.stderr.write(msg + os.linesep)
 
+    def mark_tests(self):
+        tests_map = (
```
Comment (timgraham) [inline django/db/backends/base/creation.py:36]: It might work to reuse the code but currently there's the drawback that a "skip reason" isn't set.

**Review by timgraham — COMMENTED** [review]
This is on the right track. See also my WIP at the first commit of https://github.com/cockroachdb/django-cockroachdb/pull/194. It's not highly polished.

**Comment by timgraham** [comment]
Looking good. One problem I found while working on my implementation is that I'm not sure if `@cached_property` supports inheritance. The GIS backends, for example, may need to add some skips like `geogapp/tests.py:    @skipIf(spatialite, "SpatiaLite doesn't support distance lookups with Distance objects.")`.

**Comment by hramezani** [comment]
@timgraham It seems `@cached_property` works fine. I override the `skip_tests` in `backends/spatialite/features.py ` and remove `skipIf` from `test02_distance_lookup` in `geogapp/tests.py` and it works as expected

**Review by timgraham — COMMENTED** [review]
You could do a careful audit of `connection.vendor` usage in tests and see if any more expected failures / skips could be added. I found a few at first glance:
https://github.com/django/django/blob/b9fd09d21b722c3745501151f9cbd5b95ccc3177/tests/annotations/tests.py#L23-L33
https://github.com/django/django/blob/b9fd09d21b722c3745501151f9cbd5b95ccc3177/tests/aggregation/tests.py#L1221-L1225

Reflecting on the second example, I wonder if we should move what's currently as comments in`skip_tests` to the datastructure so we can use it as the skip reason. That would be quite a bit more helpful than just "unsupported by <database>" which would require having to look up the test in `features.py` and finding the comment above the test.

I'm not in love with this but something like:
```python
skip_tests = {
    'GROUP BY optimization does not work properly when...' : {
        'aggreation.tests.AggregateTestCase.test_aggregation_subquery_annotation_multivalued',
        ...
    },
    # in other words
    'skip reason': {
        <set of tests>
    }
    ...
}
```

File: django/db/backends/base/creation.py [hunk 3]
```diff
@@ -47,6 +71,9 @@ def create_test_db(self, verbosity=1, autoclobber=False, serialize=True, keepdb=
                 self._get_database_display_str(verbosity, test_database_name),
             ))
 
+        if os.environ.get('DJANGO_TEST_SUITE') == '1':
```
[the remark this answers is not in the document]
Comment (hramezani) [reply django/db/backends/base/creation.py:74]: Do you mean that we should move the `connection.creation.mark_expected_failures_and_skips()` under this block in `runtests.py`?

https://github.com/django/django/blob/55fabc53373a8c7ef31d8c4cffd2a07be0a88c2e/tests/runtests.py#L299-L300

[the remark this answers is not in the document]
Comment (timgraham) [reply django/db/backends/base/creation.py:74]: I'm not sure where exactly in runtests.py (if you have trouble with it, I'll look it at, hopefully it's feasible) but no, not there. The logic should run regardless of what test runner is used. The idea is that `runtests.py` is used for running Django's tests suite. That's all we care about. The test runner that's used doesn't matter. Jenkins, for example, uses  `TEST_RUNNER = 'xmlrunner.extra.djangotestrunner.XMLTestRunner'`. We still need the logic to run in that case.

File: tests/backends/base/test_creation.py [hunk 4]
```diff
@@ -107,6 +108,18 @@ def test_migrate_test_setting_true(self, mocked_sync_apps, mocked_migrate, *mock
             with mock.patch.object(creation, '_destroy_test_db'):
                 creation.destroy_test_db(old_database_name, verbosity=0)
 
+    @mock.patch.dict(os.environ, {"DJANGO_TEST_SUITE": "0"})
```
[the remark this answers is not in the document]
Comment (hramezani) [reply tests/backends/base/test_creation.py]: I removed this test

File: django/db/backends/base/creation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (hramezani) [reply django/db/backends/base/creation.py]: I moved the `connection.creation.mark_expected_failures_and_skips()` to `tests.runtests.django_tests` function.
It works with django test runner and `django-docker-box`.

**Comment by hramezani** [comment]
Thanks @timgraham for the review.
I found more tests that should be skipped and moved them to `django_test_skips`.

I tried to move `test_closes_connections`
https://github.com/django/django/blob/adb40d217ec57ade46b1394cfbf3c513dc669445/tests/servers/test_liveserverthread.py#L14-L18

But, the settings dict is not initialized at the beginning of the test running, and `self.connection.is_in_memory_db()` returns `False` when I moved the test to `sqlite3.features.django_test_skips`

**Comment by hramezani** [comment]
@timgraham Could you please take another look? please let me know if I missed something. Thanks

**Comment by timgraham** [comment]
buildbot, test on oracle.

**Comment by timgraham** [comment]
Thanks. I pushed some mostly cosmetic edits. The issue with `test_closes_connections` might indicate that `mark_expected_failures_and_skips()` should be called later in the testing process (which probably means it couldn't happen in `runtests.py`). At this point, I think we can have someone else take a look and give input.

**Comment by hramezani** [comment]
Thanks @timgraham. I unchecked `Patch needs improvement`

File: django/db/backends/oracle/features.py [hunk 5]
```diff
@@ -72,6 +72,28 @@ class DatabaseFeatures(BaseDatabaseFeatures):
...
+            'db_functions.text.test_sha224.SHA224Tests.test_transform',
+        },
+        "Oracle doesn't support bitwise XOR.": {
+            'expressions.tests.ExpressionOperatorTests.test_lefthand_bitwise_xor',
+            'expressions.tests.ExpressionOperatorTests.test_lefthand_bitwise_xor_null',
+        },
+        "Oracle requires ORDER BY in row_number, ANSI:SQL doesn't.": {
+            'expressions_window.tests.WindowFunctionTests.test_row_number_no_ordering',
+        },
+        "Raises ORA-00600: internal error code on Oracle 18.": {
```
Comment (felixxm) [inline django/db/backends/oracle/features.py:87]: Single quotes.

File: django/db/backends/mysql/features.py [hunk 6]
```diff
@@ -50,6 +50,52 @@ class DatabaseFeatures(BaseDatabaseFeatures):
...
+                'https://jira.mariadb.org/browse/MDEV-22775': {
+                    'schema.tests.SchemaTests.test_alter_pk_with_self_referential_field',
+                },
+            })
+        if not self.supports_explain_analyze:
+            skips.update({
+                'MariaDB and MySQL >= 8.0.18 specific.': {
+                    'queries.test_explain.ExplainTests.test_mysql_analyze',
+                },
+            })
```
Comment (felixxm) [inline django/db/backends/mysql/features.py:96]: We cannot move these skips, because that will cause running queries before a test database is created, see 493b26bbfc9cc0ad223ece131741cba2312ced0f (ticket-31888).

File: django/contrib/gis/db/backends/mysql/features.py [hunk 7]
```diff
@@ -24,3 +24,14 @@ def supports_empty_geometry_collection(self):
...
+
+    @cached_property
+    def django_test_skips(self):
+        skips = super().django_test_skips
+        if not self.connection.mysql_is_mariadb and self.connection.mysql_version < (8, 0, 0):
+            skips.update({
+                'MySQL < 8 gives different results.': {
+                    'gis_tests.geoapp.tests.GeoLookupTest.test_disjoint_lookup',
+                },
+            })
```
Comment (felixxm) [inline django/contrib/gis/db/backends/mysql/features.py:36]: This must be an inner skip, please revert (ticket-31888).

File: tests/backends/base/test_creation.py [hunk 8]
```diff
@@ -188,3 +188,45 @@ def test_serialize_db_to_string_base_manager(self):
             data = connection.creation.serialize_db_to_string()
         self.assertIn('"model": "backends.schoolclass"', data)
         self.assertIn('"year": 1000', data)
+
```
Comment (felixxm) [inline tests/backends/base/test_creation.py:191]: I'm not sure if it's the best place for these tests :thinking:, maybe `backends/base/test_base.py`.

File: tests/backends/base/test_creation.py [hunk 9]
```diff
@@ -188,3 +188,45 @@ def test_serialize_db_to_string_base_manager(self):
...
+        creation.connection.features.django_test_skips = {
+            'skip test class': {
+                'backends.base.test_creation.SkipTestClass',
+            },
+            'skip test function': {
+                'backends.base.test_creation.skip_test_function',
+            },
+        }
+        creation.mark_expected_failures_and_skips()
+        self.assertTrue(expected_failure_test_function.__unittest_expecting_failure__)
```
Comment (felixxm) [inline tests/backends/base/test_creation.py:222]:
```suggestion
        self.assertIs(expected_failure_test_function.__unittest_expecting_failure__, True)
```

**Review by felixxm — COMMENTED** [review]
Thanks both :rocket: I left comments.

**Comment by felixxm** [comment]
Also, I think we need a release notes.

File: tests/backends/base/test_creation.py (hunk 8 above)
Comment (timgraham) [reply tests/backends/base/test_creation.py:191]: It's here because `mark_expected_failures_and_skips()` is on `DatabaseCreation` (`creation.py`). `test_base.py` is for `DatabaseWrapper` (`base.py`).

File: django/db/backends/mysql/features.py (hunk 6 above)
Comment (timgraham) [reply django/db/backends/mysql/features.py:96]: Based on this and other cases, I think we should instead try to mark skips after the database is created (https://github.com/django/django/pull/13717#issuecomment-739609140).

**Comment by timgraham** [comment]
I understand release notes could be helpful. On the other hand, it seems odd to mention backend features in release notes without formally documenting them elsewhere.

File: django/db/backends/mysql/features.py (hunk 6 above)
Comment (hramezani) [reply django/db/backends/mysql/features.py:96]: I moved `mark_expected_failures_and_skips` function call to the end of `db.backends.base.creation.create_test_db`.
Also, I moved `servers.test_liveserverthread.LiveServerThreadTest.test_closes_connections` to skip tests of sqlite3

Comment (timgraham) [reply django/db/backends/mysql/features.py:96]: Assuming this works, don't forget that we'll need some solution (like the environment variable used before) to only run `mark_expected_failures_and_skips()` when running Django's test suite.

**Comment by felixxm** [comment]
> I understand release notes could be helpful. On the other hand, it seems odd to mention backend features in release notes without formally documenting them elsewhere.

True, but I think it's still valuable for maintainers of 3rd-party database backends :thinking:

File: tests/backends/base/test_creation.py [hunk 10]
```diff
@@ -188,3 +188,45 @@ def test_serialize_db_to_string_base_manager(self):
...
+            },
+        }
+        creation.mark_expected_failures_and_skips()
+        self.assertIs(expected_failure_test_function.__unittest_expecting_failure__, True)
+        self.assertIs(SkipTestClass.__unittest_skip__, True)
+        self.assertEqual(
+            SkipTestClass.__unittest_skip_why__,
+            'skip test class',
+        )
+        self.assertTrue(skip_test_function.__unittest_skip__)
```
Comment (felixxm) [inline tests/backends/base/test_creation.py:228]:
```suggestion
        self.assertIs(skip_test_function.__unittest_skip__, True)
```

File: django/db/backends/mysql/features.py (hunk 6 above)
Comment (hramezani) [reply django/db/backends/mysql/features.py:96]: I added the environment variable

**Comment by hramezani** [comment]
@felixxm  release note added

File: docs/releases/3.2.txt [hunk 11]
```diff
@@ -451,6 +451,17 @@ Tests
   comparison against a list of string representations of objects when using the
   default value for the ``transform`` argument.
 
+* Database backends now can mark tests as expected failure by adding them to
```
Comment (timgraham) [inline docs/releases/3.2.txt:454]: Please put this in a new "Database backends" section instead.

* Third-party database backends can now skip or mark as expected failures tests in Django's test suite using the new DatabaseFeatures.django_test_skips and django_test_expected_failures attributes.

The documentation isn't needed because authors should consult the existing backends to see usage.

File: django/db/backends/base/creation.py [hunk 12]
```diff
@@ -94,7 +94,8 @@ def create_test_db(self, verbosity=1, autoclobber=False, serialize=True, keepdb=
         # Ensure a connection for the side effect of initializing the test database.
         self.connection.ensure_connection()
 
-        self.mark_expected_failures_and_skips()
+        if os.environ.get('DJANGO_TEST_SUITE') == "1":
```
Comment (timgraham) [inline django/db/backends/base/creation.py:97]: single quotes

File: tests/runtests.py [hunk 13]
```diff
@@ -248,6 +248,8 @@ def _module_match_label(module_label, label):
 
     apps.set_installed_apps(settings.INSTALLED_APPS)
 
+    os.environ['DJANGO_TEST_SUITE'] = "1"
```
Comment (timgraham) [inline tests/runtests.py:251]: single quotes + `RUNNING_DJANGOS_TEST_SUITE` could be more descriptive. Also comment `# Set an environment variable that other code may consult to see if Django's own test suite is running.`

**Comment by felixxm** [comment]
Thanks :1st_place_medal:

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: django/db/backends/sqlite3/features.py [hunk 14]
```diff
@@ -50,6 +50,38 @@ class DatabaseFeatures(BaseDatabaseFeatures):
...
+                },
+            })
+        if self.connection.is_in_memory_db():
+            skips.update({
+                "the sqlite backend's close() method is a no-op when using an "
+                "in-memory database": {
+                    'servers.test_liveserverthread.LiveServerThreadTest.test_closes_connections',
+                },
+            })
+
```
Comment (timgraham) [inline django/db/backends/sqlite3/features.py:82]: chop

File: tests/backends/base/test_creation.py [hunk 15]
```diff
@@ -107,6 +108,18 @@ def test_migrate_test_setting_true(self, mocked_sync_apps, mocked_migrate, *mock
             with mock.patch.object(creation, '_destroy_test_db'):
                 creation.destroy_test_db(old_database_name, verbosity=0)
 
+    @mock.patch.dict(os.environ, {'RUNNING_DJANGOS_TEST_SUITE': ''})
+    @mock.patch.object(BaseDatabaseCreation, 'mark_expected_failures_and_skips')
+    def test_mark_expected_failures_and_skips_call(self, mark_expected_failures_and_skips, *mocked_objects):
+        test_connection = get_connection_copy()
```
Comment (timgraham) [inline tests/backends/base/test_creation.py:114]: A docstring could be nice: "mark_expected_failures_and_skips() isn't called unless RUNNING_DJANGOS_TEST_SUITE is 'true'."

**Review by timgraham — APPROVED** [review]

**Comment by felixxm** [comment]
buildbot

File: django/db/backends/base/features.py [hunk 16]
```diff
@@ -323,6 +323,13 @@ class BaseDatabaseFeatures:
         'swedish_ci': None  # Swedish case-insensitive.
     }
 
+    # A set of dotted paths to tests in Django's test suite that are expected
+    # to fail on this database.
+    django_test_expected_failures = {}
```
Comment (timgraham) [inline django/db/backends/base/features.py:328]: Oops, this should be `set()`. None of the subclasses implementations call `super()` which perhaps isn't necessary but I want the CockroachDB backend to call `super()` to get any skips/expected failures from the PostgreSQL backend  (which the CockroachDB inherits),

Comment (hramezani) [reply django/db/backends/base/features.py:328]: I can create a PR to fix it

Comment (hramezani) [reply django/db/backends/base/features.py:328]: Here is the PR https://github.com/django/django/pull/13761

Comment (felixxm) [reply django/db/backends/base/features.py:328]: Good catch :dart:
